src/servers/http_server.cc

```diff
@@ -321,280 +142,291 @@ class HTTPAPIServer : public HTTPServerImpl {
-         endpoint_names_.end())) {
-      HandleRepository(req, rest);
-      return;
-    }
-  }
-
-  LOG_VERBOSE(1) << "HTTP error: " << req->method << " " << req->uri->path->full
-                 << " - " << static_cast<int>(EVHTP_RES_BADREQ);
-  evhtp_send_reply(req, EVHTP_RES_BADREQ);
+  SSL_load_error_strings();
+  SSL_library_init();
+  OpenSSL_add_all_algorithms();
+
+  accept_ctx.ssl_ctx = SSL_CTX_new(SSLv23_server_method());
+  SSL_CTX_set_options(accept_ctx.ssl_ctx, SSL_OP_NO_SSLv2);
+
+#ifdef SSL_CTX_set_ecdh_auto
+  SSL_CTX_set_ecdh_auto(accept_ctx.ssl_ctx, 1);
+#endif
+
+  /* load certificate and private key */
+  if (SSL_CTX_use_certificate_chain_file(accept_ctx.ssl_ctx, cert_file) != 1) {
+    fprintf(
+        stderr,
+        "an error occurred while trying to load server certificate file:%s\n",
+        cert_file);
+    return -1;
+  }
+  if (SSL_CTX_use_PrivateKey_file(
+          accept_ctx.ssl_ctx, key_file, SSL_FILETYPE_PEM) != 1) {
+    fprintf(
+        stderr, "an error occurred while trying to load private key file:%s\n",
+        key_file);
+    return -1;
+  }
+
+  if (SSL_CTX_set_cipher_list(accept_ctx.ssl_ctx, ciphers) != 1) {
+    fprintf(stderr, "ciphers could not be set: %s\n", ciphers);
+    return -1;
+  }
+
+  /* setup protocol negotiation methods */
+#if H2O_USE_NPN
+  h2o_ssl_register_npn_protocols(accept_ctx.ssl_ctx, h2o_http2_npn_protocols);
+#endif
+#if H2O_USE_ALPN
+  h2o_ssl_register_alpn_protocols(accept_ctx.ssl_ctx, h2o_http2_alpn_protocols);
+#endif
+
+  return 0;
 }
 
-void
-HTTPAPIServer::HandleHealth(evhtp_request_t* req, const std::string& health_uri)
+h2o_pathconf_t*
+HTTPAPIServer::register_handler(
+    h2o_hostconf_t* hostconf, const char* path,
+    int (*on_req)(h2o_handler_t*, h2o_req_t*))
 {
-  if (req->method != htp_method_GET) {
-    evhtp_send_reply(req, EVHTP_RES_METHNALLOWED);
-    return;
-  }
+  h2o_pathconf_t* pathconf = h2o_config_register_path(hostconf, path, 0);
+  h2o_custom_req_handler_t* handler =
+      reinterpret_cast<h2o_custom_req_handler_t*>(
+          h2o_create_handler(pathconf, sizeof(*handler)));
+  handler->http_server = this;
+  handler->super.on_req = on_req;
+  return pathconf;
+}
 
+int
+HTTPAPIServer::Health(h2o_handler_t* handler_self, h2o_req_t* req)
+{
+  h2o_custom_req_handler_t* self = (h2o_custom_req_handler_t*)handler_self;
+  if (!h2o_memis(req->method.base, req->method.len, H2O_STRLIT("GET"))) {
+    req->res.status = 400;
+    req->res.reason = "Only GET method is allowed";
+    h2o_add_header(
+        &req->pool, &req->res.headers, H2O_TOKEN_CONTENT_TYPE, NULL,
+        H2O_STRLIT("text/plain"));
+    h2o_generator_t generator = {NULL, NULL};
+    h2o_iovec_t body = h2o_strdup(&req->pool, "", SIZE_MAX);
+    h2o_start_response(req, &generator);
+    h2o_send(req, &body, 1, H2O_SEND_STATE_FINAL);
+    return 0;
+  }
+
+  re2::RE2 health_regex_(R"(/api/health/(live|ready))");
```

Comment:
Instead of this "second level" regex can we just register both paths and leave it to h2o to do the correct routing

Reply:
I will verify but to the best of my understanding h2o just accepts string paths and uses them to check if the url starts with that path. I will read some more and see if it is possible to use a regex instead.